### PR TITLE
RevealDialog and minor update

### PR DESCRIPTION
**Pull request by asolis90**
@roideuniverse  @BrandonRomano

### CODE REVIEW

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java [hunk 1]
```diff
@@ -40,15 +37,14 @@
...
+            explode.setDuration(duration);
+            explode.addListener(listener);
+            return explode;
         }
-
-        Explode explode = new Explode();
-        explode.setDuration(duration);
-        explode.addListener(listener);
-        return explode;
+        return null;
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:47]: why not add a Link check to remind the user that it does not work below `LOLLIPOP` ?
Have concern about returning null

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java [hunk 2]
```diff
@@ -62,15 +58,14 @@ public static Transition getExplodeTransition(Transition.TransitionListener list
     public static Transition getSlideTransition(Transition.TransitionListener listener, int
             duration, int gravity)
     {
-        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP)
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:61]: Instead of us doing a check, is it possible to leave it to the user of this lib to add a check ?

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java (hunk 1 above)
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:47]: What about users that are using api version 19 and above ?

Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:47]: Lint Check will only remove warnings in code

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java (hunk 2 above)
Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:61]: I think that should be a check the library has to do to make sure everything is running as expected..

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java [hunk 3]
```diff
@@ -107,30 +100,29 @@ public static void circularReveal(final View view, Context context, RevealGravit
...
+                y = 0;
                 break;
             case TOP_RIGHT:
-                x = view.getRight();
-                y = view.getTop();
+                x = view.getWidth();
+                y = 0;
                 break;
         }
+        Log.d("x,y", "" + x + " ," + y);
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java]: I think we no longer need this log :smile_cat:

**Comment by roideuniverse** [comment]
:+1:  From my side!

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java [hunk 4]
```diff
@@ -196,14 +170,14 @@ public static Point getViewEpicenter(View view)
      */
     public static Transition getFadeTransition(Transition.TransitionListener listener, int fadeMode)
     {
-        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP)
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:173]: Line 169, Could be null ? I think we should write when it will be null :cat:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        isReveal = false;
+    }
+
+
+    @Override
+    public boolean onTouchEvent(MotionEvent event)
+    {
+        if(event.getAction() == MotionEvent.ACTION_DOWN && ! isAnimating)
+        {
+            if(hasTarget)
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: What is the target ?

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+            {
+                contractDialogWithTarget();
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:51]: If this `contract` is the one like the one in `expand` and `contract`, I think we could using hide/show/reveal/expand ?

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+            getWindow().setBackgroundDrawable(new ColorDrawable(android.graphics.Color
+                    .TRANSPARENT));
+            setOnShowListener(new OnShowListener()
+            {
+                @Override
+                public void onShow(DialogInterface sdialog)
+                {
+                    if(hasTarget)
+                    {
+                        hide();
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:132]: probably we can rename `hide` to reflect what it is `hiding` :smile_cat:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 5 above)
Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: Target is the view that you want the dialog to be animated with.. In this case the user sets the target he wants to animate with dialog.. if there is no target view then it will just do a simple reveal

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+        if(isAnimating)
+        {
+            return;
+        }
+
+        isAnimating = true;
+
+        // get display metrics
+        DisplayMetrics metrics = new DisplayMetrics();
+        getWindow().getWindowManager().getDefaultDisplay().getMetrics(metrics);
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: we can simply use `context.getResources().getDisplayMetrics()` ?

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 7 above)
Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:132]: hide() is directly related to the Dialog class itself..

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 8 above)
Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: :+1:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+        targetView.setEnabled(false);
+        targetView.setClickable(false);
+
+        // change alpha of contentView and visibility
+        contentView.setAlpha(0f);
+        contentView.setVisibility(View.VISIBLE);
+
+        // get original position of targetView
+        final int originalPos[] = new int[2];
+        targetView.getLocationOnScreen(originalPos);
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: getLocationOnScreen has been used multiple times in the code, probably it will be easier to write a util function that returns an int array with the locations.
`public static int[]  getLocationOnScreen(View view)`

Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: that would but just reduce one line in the code though at multiple places :smile_cat:

Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: :+1:

Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: Good point :100:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+    }
+
+    /**
+     * Reverse reveal dialog
+     */
+    public void contractDialog()
+    {
+        int x = contentView.getWidth() / 2;
+        int y = contentView.getHeight() / 2;
+        DisplayMetrics displayMetrics = getContext().getResources().getDisplayMetrics();
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:95]: DisplayMetrics has also been used at multiple places, I think we can also keep it as a call level variable, as it is not something that will changed.

Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:95]: Yup :+1:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 7 above)
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java:132]: damn :stuck_out_tongue:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 5 above)
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: gotcha, then you don't need a boolean variable, just create a private function `boolean hasTargetView()`

**Comment by roideuniverse** [comment]
I think it would be good idea in future to propose changes of library and example application in different PR's. Kind of makes it difficult to evaluate.

**Comment by asolis90** [comment]
Gotcha.. :+1:

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+        final int yDest = (metrics.heightPixels / 2) - (targetView.getMeasuredHeight() / 2);
+
+        // animate targetView Button x transition
+        Animator targetSlideXAnim = ObjectAnimator.ofPropertyValuesHolder(targetView,
+                PropertyValuesHolder.ofFloat("translationX", 0f, xDest -
+                        originalPos[0]));
+        targetSlideXAnim.setDuration(duration);
+
+        // animate targetView y transition
+        Animator targetSlideYAnim = ObjectAnimator.ofPropertyValuesHolder(targetView,
```
Comment (roideuniverse) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: Will be a good idea to also create functions to return X and Y Property animators ?

**Comment by roideuniverse** [comment]
Its a little dense but :+1: from my side.

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java (hunk 1 above)
Comment (BrandonRomano) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/AnimationArsenal.java:47]: I think you should still add `@TargetApi(Build.VERSION_CODES.LOLLIPOP)` to the top of this method :smile_cat:.

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 5 above)
Comment (BrandonRomano) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: Or simply, `if(mTargetView != null)`.  I definitely disagree with this `hasTarget`, as it could get out of sync with the actual state (if we introduce a bug)

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+ */
+public class RevealDialog extends Dialog
+{
+
+    private int duration = 500;
+    private View mContentView;
+    private View mTargetView;
+    private boolean hasTarget;
+    private boolean isReveal;
+    private boolean isAnimating;
```
Comment (BrandonRomano) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: You should prefix these members with `m`

File: app/build.gradle [hunk 13]
```diff
@@ -22,6 +22,10 @@ android {
 dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
     testCompile 'junit:junit:4.12'
+    compile project(':android-animation-arsenal')
     compile 'com.android.support:appcompat-v7:23.1.0'
     compile 'com.android.support:design:23.1.0'
+    compile 'com.android.support:support-annotations:23.1.0'
+    // View Injection
+    compile 'com.jakewharton:butterknife:7.0.1'
```
Comment (BrandonRomano) [inline app/build.gradle:30]: I actually would advise against adding libraries if at all possible when building other libraries.

This butterknife library is simply syntactic sugar, and we could totally avoid using this.

File: app/src/main/java/com/android_animation_arsenal_example/MainActivity.java [hunk 14]
```diff
@@ -1,35 +1,233 @@
...
 
 public class MainActivity extends AppCompatActivity
 {
+    @Bind(R.id.circularReveal_iv) ImageView mImageViewReveal;
+    @Bind(R.id.specificResource_iv) ImageView mImageViewSpecified;
+    @Bind(R.id.fadeInResource_iv) ImageView mImageViewFadeIn;
+    @Bind(R.id.fadeOutResource_iv) ImageView mImageViewFadeOut;
+    @Bind(R.id.scaleResource_iv) ImageView mImageViewScale;
+    @Bind(R.id.slideLeftResource_iv) ImageView mImageViewSlideLeft;
+    @Bind(R.id.slideRightResource_iv) ImageView mImageViewSlideRight;
```
Comment (BrandonRomano) [inline app/src/main/java/com/android_animation_arsenal_example/MainActivity.java]: :eyes: `_iv`-- @roideuniverse @asolis90 I remember we had a discussion about this postfixing of the IDs before... What was the output of that conversation again?

**Comment by BrandonRomano** [comment]
I had a few comments, but this PR is actually a little too massive for me to give a really effective review without wasting a ton of time.

Check out my comments, and feel free to make changes if you agree with them.

I am going to give this a :+1: though so we can get this in.  (Read the comments before merging though)

File: app/build.gradle (hunk 13 above)
Comment (BrandonRomano) [inline app/build.gradle:30]: (totally don't remove this in this PR, if you decide you want to remove it, do it in another PR)

Comment (asolis90) [inline app/build.gradle:30]: Its mainly being used for the example app..

File: android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java (hunk 12 above)
Comment (asolis90) [inline android-animation-arsenal/src/main/java/com/android_animation_arsenal/views/RevealDialog.java]: :+1:
